File: src/vehicle_sim/src/include/PIDControl.cpp
#include "PIDControl.hpp"

PIDController::PIDController(double kp_, double ki_, double kd_, double dt_, double windup_limit_)
    : kp(kp_), ki(ki_), kd(kd_), dt(dt_), prev_error(0.0), integral(0.0), windup_limit(windup_limit_) {}

double PIDController::compute(double setpoint, double measured_value)
{
    double error = setpoint - measured_value;

    integral += error * dt;

    // apply anti-windup by clamping the integral term
    if (windup_limit > 0.0)
    {
        if (integral > windup_limit)
        {
            integral = windup_limit;
        }
        else if (integral < -windup_limit)
        {
            integral = -windup_limit;
        }
    }

    double derivative = (error - prev_error) / dt;
    prev_error = error;

    return kp * error + ki * integral + kd * derivative;
}

void PIDController::reset()
{
    prev_error = 0.0;
    integral = 0.0;
}

void PIDController::setWindupLimit(double limit)
{
    windup_limit = limit;
}

File: matlab/fcn/PIDControl.m
classdef PIDControl
    properties
        kp
        ki
        kd
        dt
        windup_limit
        integral
        prev_error
    end
    methods
        function obj = PIDControl(kp, ki, kd, dt, windup_limit)
            obj.kp = kp;
            obj.ki = ki;
            obj.kd = kd;
            obj.dt = dt;
            obj.windup_limit = windup_limit;
            obj.integral = 0;
            obj.prev_error = 0;
        end

        function c = compute(obj, setpoint, measured_value)
            error = setpoint - measured_value;

            obj.integral = obj.integral + error * obj.dt;
            obj.integral = min(max(obj.integral, -obj.windup_limit), obj.windup_limit);

            derivative = (error - obj.prev_error) / obj.dt;
            obj.prev_error = error;

            c = obj.kp * error + obj.ki * obj.integral + obj.kd * derivative;
        end

        function obj = reset(obj)
            obj.prev_error = 0;
            obj.integral = 0;
        end

        function obj = setWindupLimit(obj, limit)
            obj.windup_limit = limit;
        end
    end
end